From a2ebd88147a5cb263217863ba43eb124807a6886 Mon Sep 17 00:00:00 2001
From: Volodymyr Bendiuga <volodymyr.bendiuga@westermo.se>
Date: Wed, 23 Mar 2016 14:41:53 +0100
Subject: [PATCH 13/38] route: allow setting index for TC actions
Organization: Westermo Network Technologies AB

API:
	rtnl_tc_set_act_index()
	rtnl_tc_get_act_index()

Signed-off-by: Volodymyr Bendiuga <volodymyr.bendiuga@westermo.se>
Signed-off-by: Joachim Nilsson <troglobit@gmail.com>
---
 include/netlink-private/tc.h    |  3 ++-
 include/netlink-private/types.h |  3 ++-
 include/netlink/route/tc.h      |  2 ++
 lib/route/tc.c                  | 11 +++++++++++
 libnl-route-3.sym               |  2 ++
 5 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/include/netlink-private/tc.h b/include/netlink-private/tc.h
index b0f9c50..91cfc01 100644
--- a/include/netlink-private/tc.h
+++ b/include/netlink-private/tc.h
@@ -33,7 +33,8 @@ extern "C" {
 #define TCA_ATTR_OVERHEAD	0x1000
 #define TCA_ATTR_LINKTYPE	0x2000
 #define TCA_ATTR_CHAIN          0x4000
-#define TCA_ATTR_MAX            TCA_ATTR_CHAIN
+#define TCA_ATTR_ACT_INDEX      0x8000
+#define TCA_ATTR_MAX            TCA_ATTR_ACT_INDEX
 
 extern int tca_parse(struct nlattr **, int, struct rtnl_tc *,
                      const struct nla_policy *);
diff --git a/include/netlink-private/types.h b/include/netlink-private/types.h
index 40ba91f..bf04cdd 100644
--- a/include/netlink-private/types.h
+++ b/include/netlink-private/types.h
@@ -586,7 +586,8 @@ struct rtnl_tstats
 	struct rtnl_link *	pre ##_link;		\
 	struct rtnl_tc_ops *	pre ##_ops;		\
 	enum rtnl_tc_type	pre ##_type;		\
-	uint32_t		pre ##_chain
+	uint32_t		pre ##_chain;		\
+	uint32_t		pre ##_act_index
 
 struct rtnl_tc
 {
diff --git a/include/netlink/route/tc.h b/include/netlink/route/tc.h
index 51d670a..cf9c391 100644
--- a/include/netlink/route/tc.h
+++ b/include/netlink/route/tc.h
@@ -97,6 +97,8 @@ extern uint32_t		rtnl_tc_get_parent(struct rtnl_tc *);
 extern int		rtnl_tc_set_kind(struct rtnl_tc *, const char *);
 extern char *		rtnl_tc_get_kind(struct rtnl_tc *);
 extern uint64_t		rtnl_tc_get_stat(struct rtnl_tc *, enum rtnl_tc_stat);
+extern void             rtnl_tc_set_act_index(struct rtnl_tc *tc, int index);
+extern int              rtnl_tc_get_act_index(struct rtnl_tc *tc);
 extern char *		rtnl_tc_stat2str(enum rtnl_tc_stat, char *, size_t);
 extern int		rtnl_tc_str2stat(const char *);
 
diff --git a/lib/route/tc.c b/lib/route/tc.c
index 35303f5..b9a8336 100644
--- a/lib/route/tc.c
+++ b/lib/route/tc.c
@@ -603,6 +603,17 @@ int rtnl_tc_get_chain(struct rtnl_tc *tc, uint32_t *out_value)
 	return 0;
 }
 
+void rtnl_tc_set_act_index(struct rtnl_tc *tc, int index)
+{
+	tc->tc_act_index = index;
+	tc->ce_mask |= TCA_ATTR_ACT_INDEX;
+}
+
+int rtnl_tc_get_act_index(struct rtnl_tc *tc)
+{
+	return tc->tc_act_index;
+}
+
 /** @} */
 
 /**
diff --git a/libnl-route-3.sym b/libnl-route-3.sym
index fee2387..5b617ee 100644
--- a/libnl-route-3.sym
+++ b/libnl-route-3.sym
@@ -1228,4 +1228,6 @@ global:
 	rtnl_vlan_set_vlan_id;
 	rtnl_vlan_set_vlan_prio;
 	rtnl_route_lookup;
+	rtnl_tc_set_act_index;
+	rtnl_tc_get_act_index;
 } libnl_3_4;
-- 
2.25.1

